contracts/RMRK/extension/tokenProperties/RMRKTokenProperties.sol

```diff
@@ -0,0 +1,144 @@
+// SPDX-License-Identifier: Apache-2.0
+
+pragma solidity ^0.8.16;
+
+import "./IRMRKTokenProperties.sol";
+
+/**
+ * @title RMRKTokenProperties
+ * @author RMRK team
+ * @notice Smart contract of the RMRK Token properties module.
+ */
+contract RMRKTokenProperties is IRMRKTokenProperties {
+
+    uint256 private totalStringProperties;
+    mapping(uint256 => string) stringProperties;
+    mapping(uint256 => mapping(string => uint256)) private _stringProperties;
+
+    uint256 private totalAddressProperties;
+    mapping(uint256 => address) addressProperties;
+    mapping(uint256 => mapping(string => uint256)) private _addressProperties;
```

Comment:
I just saw this was Jan's  proposal, call to defend it @ThunderDeliverer :sweat_smile:

Reply:
String is easy, it's variable length and can consume a looot of resources in order to fully set it. You'd probably be burning through gas like your life depends on it.

Having this approach to strings allows for reuse of the same value on multiple occasions (which is quire common for large generative collections).

Address and fixed size bytes don't benefit as much (and one might argue that using an additional mapping might be more expensive), but I still feel for the continuity sake this is a good approach.

I'm open on rolling this back for addresses and bytes, but am confident this is the best approach for the strings.

Reply:
Ok I see the point now of mapping string keys to indices/ids. So you should only add a new index/id if that key does not exist already, right? 
I think we're missing that check.

Now I don't understand why you say you're convinced this is the best approach only for string values, if this affects specifically the key which is a string in every case. So shouldn't we actually do that for all properties, including uint and bool?

Reply:
Well, for uint, you are using an uint to fetch an uint (possibly of the same value), so might be more expensive than just having a direct mapping. For boolean values it's kind of the same thing, you should only need 2 or 3 (in case you consider ID of 0 as unset) values, which seems like more storage usage than just assigning it directly 🤔

Reply:
This is even better, yes! instead of using string keys in every mapping, we just use them to get the intended IDs this makes the interface easier to use and understand, but keeps the cost as low as possible.

I'd use the pattern you are proposing for all of the values other than string. For those I'd still use another mapping, so one that uses indices to retrieve the string property values:
`mapping(uint256 => string) stringProperties;`
I still feel like this might be a great improvement:
`mapping(string => id) addressKeysToIds;`
And then the mapping would be:
`mapping(uint256 => mapping(uint256 => uint256)) private _stringValues;`

So to retrieve a string value:

```
id = addressKeysToIds;
index = _stringValues[tokenId][id];
return stringProperties[index];
```

This should be easy to justify and noone should have a problem with accepting a slightly different approach.

Reply:
I don't see how this is better for strings.
Also, let's say you have some key K assigned to id A (We agree K->A will be a thing for ever now).
If you 2 tokens use this key, I think you'd be forcing them to always have the same value. Unless what I'm assuming we agree is false.

Reply:
Much like you could modify the value of `_addressValues`, you could do the same for `_stringValue`, just at a different point.

You change the `_addressValues` like this:
`_addressValues[tokenId][id] = newAddressValue;`

And you would change the string value like this:
`stringProperties[index] = newStringValue;`

It's just a different place where you can modify the value, but it is still modifiable if needed, or if you want to use another, already existing, string value, you just do:
`_stringValues[tokenId][id] = newIndex;`

Reply:
Then you would need to create a new id for the same key. Which seems bad IMO. And I still don't see what you gain for strings.

Reply:
ID could remain the same only the string value accessible by the index or the index itself would be changed if you wish to do this. I don't see the need to change IDs 🤔